File: 프로그래머스/3/150366. 표 병합/표 병합.cpp
#include <string>
#include <vector>
#include <sstream>
#include <iostream>
using namespace std;

struct xy{
    int x;
    int y;
    bool operator ==(xy a){
        return (x == a.x && y == a.y);
    }
};

xy link[51][51];
string map[51][51];
vector<string> solution(vector<string> commands) {
    vector<string> answer;

    for(int i = 1; i<= 50; i++)
        for(int j = 1; j<= 50; j++){
            link[i][j].x = i;
            link[i][j].y = j;
        }
    for(auto com : commands){
        string a,b,c,d,e;
        stringstream ss(com);
        
        ss >> a >> b >> c >> d >> e;
        
        if(a == "UPDATE"){
            if(d == ""){       
                for(int i = 1; i <= 50; i++)
                    for(int j = 1; j<=50; j++)
                        if(map[i][j] == b)
                            map[i][j] = c;               
            }
            else{
                int x = stoi(b);
                int y = stoi(c);
                map[x][y] = d;
                xy par = link[x][y];
                
                for(int i = 1; i<=50; i++)
                    for(int j = 1; j<=50; j++){
                        if(link[i][j] == par)
                            map[i][j] = d;
                    }
                
            }
        }
        else if(a == "MERGE"){
            int r1 = stoi(b);
            int c1 = stoi(c);
            int r2 = stoi(d);
            int c2 = stoi(e);
            
            string new_val = map[r1][c1];
            if(new_val == "")
                new_val = map[r2][c2];
            
            xy par1 = link[r1][c1];
            xy par2 = link[r2][c2];
            
            for(int i = 1; i<=50; i++)
                for(int j = 1; j<=50; j++){
                    if(link[i][j] == par2){
                        link[i][j] = par1;
                        map[i][j] = new_val;
                    }
                    else if(link[i][j] == par1)
                        map[i][j] = new_val;
                }
            
           
        }
        else if(a == "UNMERGE"){
            int x = stoi(b);
            int y = stoi(c);
            
            xy par = link[x][y];
            
            for(int i = 1; i<=50; i++)
                for(int j = 1; j<=50; j++)
                    if(par == link[i][j]){
                        link[i][j] = {i,j};
                        if(i != x || j != y)
                            map[i][j] = "";
                    }
                
            
        }
        else if(a == "PRINT"){
            int x = stoi(b);
            int y = stoi(c);
            string new_val = map[x][y];
            if(new_val == "")
                new_val = "EMPTY";
            answer.push_back(new_val);
        }
    }
    
    

    return answer;
}

File: 프로그래머스/3/150366. 표 병합/표 병합.java
import java.io.*;
import java.util.*;
class Solution {
    static int[] v;
    static String[] an;
    public String[] solution(String[] commands) {
        String[] answer = {};
        ArrayList<String> answerList = new ArrayList<>();
        
        v = new int[2501];
        an = new String[2501];
        for(int i = 0; i < 2501; i++) {
            v[i] = i;
            an[i] = "";
        }
        
        for(String com : commands){
            StringTokenizer st = new StringTokenizer(com);
            String s1 = st.nextToken();
            if(s1.equals("UPDATE")){
                String s2 = st.nextToken();
                String s3 = st.nextToken();
                
                if(st.hasMoreTokens()){
                    int x = Integer.parseInt(s2);
                    int y = Integer.parseInt(s3);
                    
                    String val = st.nextToken();
                    int parents = find((x - 1) * 50 + (y - 1));
                    an[ parents ] = val;
                    continue;
                }
                for(int i = 0; i < 2501; i++){
                    if(an[ find(i) ].equals(s2)) an[ find(i) ] = s3;
                }
                
            }else if(s1.equals("MERGE")){
                int x1 = Integer.parseInt(st.nextToken());
                int y1 = Integer.parseInt(st.nextToken());
                int x2 = Integer.parseInt(st.nextToken());
                int y2 = Integer.parseInt(st.nextToken());
                
                int a = (x1 - 1) * 50 + (y1 - 1);
                int b = (x2 - 1) * 50 + (y2 - 1);
                int parA = find(a);
                int parB = find(b);
                if(parA == parB) continue;
                
                String val = an[parA];
                if(an[parA].equals("")) val = an[parB];
                union(parA, parB);
                an[parA] = val;
                
            }else if(s1.equals("UNMERGE")){
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                
                int par = find((x - 1) * 50 + (y - 1));
                String val = an[ par ];
                
                for(int i = 0; i < 2501; i++) find(i); // 부모 업데이트
                
                for(int i = 0; i < 2501; i++){
                    if(find(i) == par){
                        an[i] = "";
                        v[i] = i;
                    }
                }
                
                an[(x - 1) * 50 + (y - 1)] = val;
            }else{
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                int par = find((x-1) * 50 + (y - 1));
                if(an[par].equals("")) answerList.add("EMPTY");
                else answerList.add(an[par]);
            }
        }
       
        answer = new String[answerList.size()];
        for(int i = 0; i < answer.length; i++)
            answer[i] = answerList.get(i);
        return answer;
    }
    
    
    static int find(int x){
        if(v[x] == x) return x;
        return v[x] = find(v[x]);
    }
    static void union(int x, int y){
        int a = find(x);
        int b = find(y);
        v[b] = a;
    }
}